### PR TITLE
[#128919787 FINISHED]Implementação da visualização do Aluno

### PR DESCRIPTION
**Pull request by lfp2**
Visualizar a situação do aluno, mostrando a média com a cor verde (acima de 7) e a cor vermelha(abaixo de 5). O link com histórico não foi realizado, pois a feature ainda não foi implementada.

---
This change is [Reviewable](https://reviewable.io/reviews/pauloborba/ta/49)

### CODE REVIEW

**Comment by pauloborba** [comment]
Reviewed 2 of 7 files at r2, 18 of 21 files at r3.
Review status: 20 of 25 files reviewed at latest revision, 7 unresolved discussions.

---

*[grails-app/controllers/ta/SituationStudentsController.groovy, line 21 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KY-0CRhR-lcygaZu-_M:-KY-0CRiLWdSZlU5K71f:bqrxbs0) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/grails-app/controllers/ta/SituationStudentsController.groovy#L21)):*
> ```Groovy
> 
> 
>     public List<EvaluationsByCriterion> listaCriterios(String aluno){
> ```

isso é uma ação ou um serviço? quem chama?

---

*[test/cucumber/SituationStudents.feature, line 17 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KXzychvs0w78LXQ0-UB:-KXzychvs0w78LXQ0-UC:bcww1vw) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/SituationStudents.feature#L17)):*
> ```Cucumber
>     #GUI
>     Scenario: Visualizar media positiva dos alunos
>       Given o aluno "João Vasconcelos", com login "jvsn", possui média "9" em "Requisitos de Sistemas"
> ```

não há médias numéricas associadas às metas, apenas MA, MPA e MANA

---

*[test/cucumber/SituationStudents.feature, line 30 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KXzz3_L8uVjwgHIKQKj:-KXzz3_L8uVjwgHIKQKk:b-578dit) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/SituationStudents.feature#L30)):*
> ```Cucumber
>     #GUI
>     #Scenario: Alunos repetentes na pagina de visualização
>      # Given eu estou na página "Home"
> ```

se não estivesse em home não mudaria o comportamento avaliado pelo teste, desnecessário

---

*[test/cucumber/SituationStudents.feature, line 31 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KXzzHiVZsdo4y_eIafK:-KXzzHiVZsdo4y_eIafL:b-2lpss2) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/SituationStudents.feature#L31)):*
> ```Cucumber
>     #Scenario: Alunos repetentes na pagina de visualização
>      # Given eu estou na página "Home"
>       #And o aluno "Ivan Cardoso", com login "isn", possui uma nota cadastrada em "2015"
> ```

teria que amarrar que está na turma 2016 e tal

---

*[test/cucumber/steps/SituationStudentsSteps.groovy, line 71 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KY--UTTqq3-8ceJUNC4:-KY--UTTqq3-8ceJUNC5:b5r0v6w) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/steps/SituationStudentsSteps.groovy#L71)):*
> ```Groovy
> 
> Given(~/^o aluno "([^"]*)", com login "([^"]*)", possui media "([^"]*)" em "([^"]*)"$/) { String aluno, String login, String media, String conceito ->
>     to AddStudentsPage
> ```

codigo duplicado, merecia extracao

---

*[test/functional/pages/SituationStudentPage.groovy, line 11 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KY--ZmjBs_wAaILd63o:-KY--ZmjBs_wAaILd63p:b-mjg6n7) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/functional/pages/SituationStudentPage.groovy#L11)):*
> ```Groovy
> 
>     static at =  {
>         title ==~ /Create property List/
> ```

internacionalização

---

*[test/functional/pages/SituationStudentPage.groovy, line 24 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KY--mkWAIxKsld-nWNH:-KY--mkWAIxKsld-nWNI:bgcae03) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/functional/pages/SituationStudentPage.groovy#L24)):*
> ```Groovy
> 
>     boolean mediaVermelha(aluno, conceito){
>         def x = aluno + "+" + conceito
> ```

codigo duplicado

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/ta/49)*

**Comment by lfp2** [comment]
Review status: 20 of 25 files reviewed at latest revision, 7 unresolved discussions.

---

*[grails-app/controllers/ta/SituationStudentsController.groovy, line 21 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KY-0CRhR-lcygaZu-_M:-KY-VvuMRQszY6joZmNV:bok7gct) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/grails-app/controllers/ta/SituationStudentsController.groovy#L21)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

isso é uma ação ou um serviço? quem chama?
</blockquote></details>

Acho que esqueci de tirar, era um serviço. Foi desnecessário no final.

---

*[test/cucumber/SituationStudents.feature, line 17 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KXzychvs0w78LXQ0-UB:-KY-VEIilZwIDU9JXpfz:b-g6r92x) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/SituationStudents.feature#L17)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

não há médias numéricas associadas às metas, apenas MA, MPA e MANA
</blockquote></details>

No caso, eu deixei a meta numérica porque o sistema não possui o cálculo com MA, MPA e MANA (Não consegui implementar este calculo a tempo)

---

*[test/cucumber/SituationStudents.feature, line 30 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KXzz3_L8uVjwgHIKQKj:-KY-VVn_tNNQCcRmPUX3:bpb10o) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/SituationStudents.feature#L30)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

se não estivesse em home não mudaria o comportamento avaliado pelo teste, desnecessário
</blockquote></details>

Irei retirar do teste então

---

*[test/cucumber/SituationStudents.feature, line 31 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KXzzHiVZsdo4y_eIafK:-KY-Vd5z7gcMwrmr78e8:bm8tqwc) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/SituationStudents.feature#L31)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

teria que amarrar que está na turma 2016 e tal
</blockquote></details>

Vou adicionar que ele está presente na turma 2016

---

*[test/cucumber/steps/SituationStudentsSteps.groovy, line 71 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KY--UTTqq3-8ceJUNC4:-KY-ViUwTHrkzfYfH2Hg:b-zig8yz) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/cucumber/steps/SituationStudentsSteps.groovy#L71)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

codigo duplicado, merecia extracao
</blockquote></details>

Ele estava dando erro sem estar repetido

---

*[test/functional/pages/SituationStudentPage.groovy, line 24 at r5](https://reviewable.io:443/reviews/pauloborba/ta/49#-KY--mkWAIxKsld-nWNH:-KY-Vq4sE86TN4EkVi-Y:bqngo7j) ([raw file](https://github.com/pauloborba/ta/blob/c69c019d89f377d9bf96463986f6eb35619a9ff9/test/functional/pages/SituationStudentPage.groovy#L24)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

codigo duplicado
</blockquote></details>

Irei  colocar no método acima

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/ta/49)*

File: test/functional/pages/SituationStudentPage.groovy [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import geb.Page
+/**
+ * Created by lavinia on 25/11/2016.
+ */
+class SituationStudentPage extends Page{
+    static url = "/TA/situationStudents/index"
+
+    static at =  {
+        title ==~ /Create property List/
```
Comment (jvsn) [inline test/functional/pages/SituationStudentPage.groovy:11]: Internacionalizar

File: test/cucumber/steps/SituationStudentsSteps.groovy [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import ta.Student
+
+
+/**
+ * Created by lavin on 29/11/2016.
+ */
+this.metaClass.mixin(cucumber.api.groovy.Hooks)
+this.metaClass.mixin(cucumber.api.groovy.EN)
+
+List<EvaluationsByCriterion> criteriosLista
```
Comment (jvsn) [inline test/cucumber/steps/SituationStudentsSteps.groovy:18]: Variável global não é uma boa prática

File: test/cucumber/SituationStudents.feature [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    As the teacher
+    I want to see the situation of student
+    So I can have a better control of students
+
+    #Controle
+    Scenario: Visualizar a nota do aluno
+      Given o aluno "João Vasconcelos", com login "jvsn", possui conceitos "MA", "MPA" e "MA" em "RS"
+      And o aluno "João Vasconcelos", possui media "8" em "RS"
+      When eu solicito a situação do aluno "João Vasconcelos"
+      Then a média de "João Vasconcelos" em "RS" continua sendo "8"
```
Comment (jvsn) [inline test/cucumber/SituationStudents.feature:13]: Existe algum caso em que a média mude? Seria mais adequado um
Then eu posso ver o aluno "João Vasconcelos" com média "8" em "RS"

File: test/cucumber/SituationStudents.feature [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      Then a média do aluno de login "jvsn" em "Requisitos de Sistemas" aparece verde e com uma seta para cima
+
+   #GUI
+    Scenario: Visualizar media negativa dos alunos
+      Given o aluno "Ivan Cardoso", com login "isn", possui media "3" em "Requisitos de Sistemas"
+      When eu solicito a página "Visualização"
+      Then a média do aluno de login "isn" em "Requisitos de Sistemas" aparece vermelho e com uma seta para baixo
+ ##
+
+    #GUI
```
Comment (jvsn) [inline test/cucumber/SituationStudents.feature:28]: Limpar testes comentados para não carregar o código

File: test/cucumber/steps/SituationStudentsSteps.groovy [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    page.selectAddEvaluation()
+}
+When(~/^eu solicito a página "([^"]*)"$/) { String pagina ->
+    if(pagina.equals("SituationStudentPage")){
+        to SituationStudentPage
+        at SituationStudentPage
+    }
+}
+Then(~/^a média do aluno de login "([^"]*)" em "([^"]*)" aparece verde e com uma seta para cima$/) { String aluno, String conceito ->
```
Comment (jvsn) [inline test/cucumber/steps/SituationStudentsSteps.groovy]: O teste explicita uma seta para cima, porém esta condição aparentemente não é testada.

File: test/cucumber/steps/SituationStudentsSteps.groovy [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    at AddStudentsPage
+    page.fillStudentDetails(aluno, login)
+    page.selectAddStudent()
+    to CreateCriterionPage
+    at CreateCriterionPage
+    page.createCriterion(conceito)
+    to AddEvaluationPage
+    at AddEvaluationPage
+    page.chooseCriterion(conceito)
+    if(media.equals("9")){
```
Comment (jvsn) [inline test/cucumber/steps/SituationStudentsSteps.groovy:55]: else? O que é escolhido quando a média não e igual a 9?

File: test/cucumber/steps/SituationStudentsSteps.groovy [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    if(pagina.equals("SituationStudentPage")){
+        to SituationStudentPage
+        at SituationStudentPage
+    }
+}
+Then(~/^a média do aluno de login "([^"]*)" em "([^"]*)" aparece verde e com uma seta para cima$/) { String aluno, String conceito ->
+        assert page.mediaVerde(aluno, conceito)
+}
+
+Given(~/^o aluno "([^"]*)", com login "([^"]*)", possui media "([^"]*)" em "([^"]*)"$/) { String aluno, String login, String media, String conceito ->
```
Comment (jvsn) [inline test/cucumber/steps/SituationStudentsSteps.groovy:70]: Este mesmo given foi explicitado anteriormente. Devido ao problema do if else existe dois Givens diferentes para o mesmo teste.

Comment (jvsn) [reply test/cucumber/steps/SituationStudentsSteps.groovy:70]: Addendo: gera problema de código duplicado

File: test/cucumber/steps/SituationStudentsSteps.groovy [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    to AddEvaluationPage
+    at AddEvaluationPage
+    page.chooseCriterion(conceito)
+    if(media.equals("3")){
+        page.chooseValue("MANA")
+    }
+    page.selectAddEvaluation()
+}
+
+Then(~/^a média do aluno de login "([^"]*)" em "([^"]*)" aparece vermelho e com uma seta para baixo$/) { String aluno, String conceito ->
```
Comment (jvsn) [inline test/cucumber/steps/SituationStudentsSteps.groovy]: Mais uma vez a questão das setas. Com essa abordagem também temos um Then com códigos semelhantes, duplicando-os. Poderiamos escrever:
`Then(~/^a média do aluno de login "([^"]*)" em "([^"]*)" aparece "([^"]*)" e com uma seta para "([^"]*)"$/) { String aluno, String conceito, String cor, String sentido ->`
Ou até mesmo retirar a complexidade da seta.

File: test/cucumber/steps/SituationStudentsSteps.groovy [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    page.createCriterion(conceito)
+    to AddEvaluationPage
+    at AddEvaluationPage
+    page.chooseCriterion(conceito)
+    if(media.equals("9")){
+        page.chooseValue("MA")
+    }
+    page.selectAddEvaluation()
+}
+When(~/^eu solicito a página "([^"]*)"$/) { String pagina ->
```
Comment (jvsn) [inline test/cucumber/steps/SituationStudentsSteps.groovy:60]: Não necessita de um if. Se for a página desejada, ela retorna true, caso contrario, retorna false.

**Review by jvsn — COMMENTED** [review]
Os testes de gui de Lavínia foram bem explicativos, testando cada passo dado. Porém houve algumas duplicações de códigos, variáveis globais. Era possível enxugar ainda mais os testes.

File: test/cucumber/steps/SituationStudentsSteps.groovy (hunk 6 above)
Comment (lfp2) [reply test/cucumber/steps/SituationStudentsSteps.groovy:55]: É só para averiguar se o que foi passado como feature foi o 9

**Pull request closed without merging**